incubator/hnc/internal/validators/hncconfig.go

```diff
@@ -76,83 +76,71 @@ func (c *HNCConfig) Handle(ctx context.Context, req admission.Request) admission
 // handle implements the validation logic of this validator for Create and Update operations,
 // allowing it to be more easily unit tested (ie without constructing a full admission.Request).
 func (c *HNCConfig) handle(ctx context.Context, inst *api.HNCConfiguration) admission.Response {
+	ts := gvkSet{}
+	// Convert all valid types from GR to GVK. If any type is invalid, e.g. not
+	// exist in the apiserver, wrong configuration, deny the request.
+	if rp := c.validateTypes(inst, ts); !rp.Allowed {
+		return rp
+	}
+
+	// Lastly, check if changing a type to "Propagate" mode would cause
+	// overwriting user-created objects.
+	return c.checkForest(inst, ts)
+}
+
+func (c *HNCConfig) validateTypes(inst *api.HNCConfiguration, ts gvkSet) admission.Response {
 	roleExist := false
 	roleBindingExist := false
-	ts := gvkSet{}
 	for _, t := range inst.Spec.Types {
-		if rp := c.isTypeConfigured(t, ts); !rp.Allowed {
-			return rp
+		// Validate the type exists in the apiserver. If yes, convert GR to GVK.
+		gr := schema.GroupResource{Group: t.Group, Resource: t.Resource}
+		gvk, err := c.validator.GVKFor(gr)
```

Comment:
Why do we need to index this on the gvk? Can't we index it on the gr?

Reply:
We will need `ts` (types with mode) to be stored as `map[gvk]mode` because we will `checkForest()` next to see if any source overwriting would happen if the mode is changed.

Reply:
Maybe add that to the comment on line 95? Say _why_ we're converting from GR to GVK as opposed to just stating that we're doing it.